src/PrettyPrinting.jl

```diff
@@ -1401,4 +1403,158 @@ function with_unicode(f::Function)
   allow_unicode(old_allow_unicode);
 end
 
+################################################################################
+#
+#  IO context with indendation and lowercasing
+#
+################################################################################
+
+import Base: convert, show, pipe_reader, pipe_writer, lock, unlock, write,
+             getindex, in, haskey, get, print, unwrapcontext
+
+export IOCustom, Indent, Dedent, indent_string!
+
+struct Indent end
+struct Dedent end
+
+struct Lowercase end
+struct Uppercase end
+
+Base.show(io::IO, ::Union{Indent, Dedent}) = nothing
+
+mutable struct IOCustom{IO_t <: IO} <: Base.AbstractPipe
+    io::IO_t
+    indent_level::Int
+    aligns::Vector{Int}
+    offset::Int
+    indented_line::Bool
+    indent_str::String
+    align_char::Char
+    printed::Int
+    lowercasefirst::Bool
+
+    function IOCustom{IO_t}(io::IO_t, indent_level::Int, aligns::Vector{Int},
+                            offset::Int, indented_line::Bool, indent_str::String,
+                            align_char::Char, printed::Int, lowercasefirst::Bool) where IO_t <: IO
+        @assert(!(IO_t <: IOCustom))
+        return new(io, indent_level, aligns, offset, indented_line, indent_str, align_char, printed, lowercasefirst)
+    end
+end
+
+_unwrap(io::IOCustom) = io
+
+_unwrap(io::IOContext) = io.io
+
+indent_string!(io::IO, str::String) = (_unwrap(io).indent_str = str; io)
+
+IOCustom(io::IO) = IOCustom{typeof(io)}(io, 0, Int[], 0, false, "  ", ' ', 0, false)
+
+IOCustom(io::IOCustom) = io
+
+convert(::Type{IOCustom}, io::IOCustom) = io
+
+in(key_value::Pair, io::IOCustom) = in(key_value, io.io, ===)
+haskey(io::IOCustom, key) = haskey(io.io, key)
+getindex(io::IOCustom, key) = getindex(io.io, key)
+get(io::IOCustom, key, default) = get(io.io, key, default)
+
+function show(_io::IO, io::IOCustom)
+    ioi = IOCustom(_io)
+    print(ioi, "IOCustom:", Indent())
+    print(ioi, "\nIO: "); show(ioi, io.io)
+    print(ioi, "\nIndent string: \"", io.indent_str, "\"")
+    print(ioi, "\nIndent: ", io.indent_level)
+end
+
+pipe_reader(io::IOCustom) = io.io
+pipe_writer(io::IOCustom) = io.io
+lock(io::IOCustom) = lock(io.io)
+unlock(io::IOCustom) = unlock(io.io)
+
+Base.displaysize(io::IOCustom) = displaysize(io.io)
+
+write(io::IO, ::Indent) = (_unwrap(io).indent_level += 1; 0)
+print(io::IO, ::Indent) = write(io, Indent())
+write(io::IO, ::Dedent) = (_unwrap(io).indent_level = max(0, io.indent_level - 1); 0)
+print(io::IO, ::Dedent) = write(io, Dedent())
+write(io::IO, ::Lowercase) = (_unwrap(io).lowercasefirst = true; 0)
+print(io::IO, ::Lowercase) = write(io, Lowercase())
+write(io::IO, ::Uppercase) = (_unwrap(io).lowercasefirst = false; 0)
+print(io::IO, ::Uppercase) = write(io, Uppercase())
```

Comment:
So `Lowercase` causes stuff afterwards to be printed with first letter in lowercase; but `Uppercase` then does not do the same for uppercase, but instead it resets the settings made by `Lowercase`. That seems confusing. Also perhaps in the future we want to be able to do the reverse, and force the first letter to be upper case (My editors calls that "Capital case" resp. "Capitalize Words")

Maybe `Reset` or `Clear` or `UnchangedCase` or `LowercaseOff` so to undo the effect ?

Reply:
And while I am at it: why does `Lowercase` map to `lowercasefirst` and not to `lowercase`?

Reply:
Because it should only change the case of the next word and not of every letter that will every be printed afterwards. For example, it would be awkward to print "Type name with Hilbert". I had the impression it might be better to only change the case of the first letter that follows.

I agree that the names can be made a bit clearer. I will go with `Lowercase()` and `LowercaseOff()` if there is no objection within the next minutes.